src/runtime/swg-client-event-manager-test.js

```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright 2018 The Subscribe with Google Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AnalyticsEvent,EventOriginator} from '../proto/api_messages';
+import {
+  SwgClientEvent, SwgClientEventManager, ShouldFilter,
+} from './swg-client-event-manager';
+
+const DEFAULT_TYPE = AnalyticsEvent.IMPRESSION_AD;
+const DEFAULT_ORIGIN = EventOriginator.SWG_CLIENT;
+const DEFAULT_EVENT = new SwgClientEvent(DEFAULT_TYPE, DEFAULT_ORIGIN);
+const OTHER_TYPE = AnalyticsEvent.ACTION_PAYMENT_COMPLETE;
+const OTHER_ORIGIN = EventOriginator.AMP_CLIENT;
+const BAD_VALUE = 'I should throw an error';
+const EventManager = new SwgClientEventManager();
+
+describes.sandboxed('SwgClientEvent', {}, () => {
+  it('should respect properties that have been set', () => {
+    const event = new SwgClientEvent(DEFAULT_TYPE, DEFAULT_ORIGIN);
+    expect(event.getEventType()).to.equal(DEFAULT_TYPE);
+    expect(event.getEventOriginator()).to.equal(DEFAULT_ORIGIN);
+    expect(event.getIsFromUserAction()).to.be.null;
+
+    event.setEventType(OTHER_TYPE);
+    event.setEventOriginator(OTHER_ORIGIN);
+    event.setIsFromUserAction(true);
+    event.setAdditionalParameters({aValue: 45});
+
+    expect(event.getEventType()).to.equal(OTHER_TYPE);
+    expect(event.getEventOriginator()).to.equal(OTHER_ORIGIN);
+    expect(event.getIsFromUserAction()).to.be.true;
+    expect(event.getAdditionalParameters().aValue).to.equal(45);
+
+    event.setIsFromUserAction(null);
+    expect(event.getIsFromUserAction()).to.be.null;
+  });
+
+  it('should have a working copy constructor', () => {
```

Comment:
Is this necessary? A copy constructor, given we are going to use a typedef for SwgClientEvent?

Reply:
I thought it would be useful for the AMP client to be able to intercept the SwG event, suppress it, copy it and modify it by just setting the originator to itself and submitting a new event.  This means that if we add properties later AMP *might* be able to avoid some updates to set the properties because it can just take the ones set by SwG.

Copy might not be necessary thought - it could suppress the event and modify it without copying it.  I'm amenable to that.

Reply:
thanks I'll resolve the conversation when i see a patch that addresses this.

Reply:
The other thing that had occurred to me is that someone could declare a const at the top of the file like

const MY_EVENT = new SwgClientEvent(...);

And pass that to event manager.  We do not want people modifying these events.  So if they want to reuse it - they should copy it and make it their own.